stock_picking_location_check/models/stock_picking.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2024 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    allow_location_inconsistency = fields.Boolean(
+        copy=False,
+        help="If enabled, no error is raised for location inconsistency between "
+        "picking and its move lines at the time of validation.",
+    )
+
+    def _get_child_location_ids(self, location):
+        return self.env["stock.location"].search(
+            [("id", "child_of", location.id), ("usage", "!=", "view")]
+        )
+
+    def _check_location_consistency(
+        self, pick_location, pick_dest_location, line_locations, line_dest_locations
+    ):
+        self.ensure_one()
+        if not (
+            set(line_locations).issubset(self._get_child_location_ids(pick_location))
+            and set(line_dest_locations).issubset(
+                self._get_child_location_ids(pick_dest_location)
+            )
+        ):
```

Comment:
I suppose `_get_child_location_ids()` doesn't work recursively.

```suggestion
    def _check_location_consistency(self, pick_location, line_locations):
        self.ensure_one()
        if not (
            set(line_locations).issubset(
                pick_location.child_internal_location_ids or pick_location
            )
        ):
```

Reply:
`child_of` works recursively. I checked with hierarchical child locations and works well.
Reference: https://stackoverflow.com/questions/59562532/explanation-of-odoo-filter-domain-operator-child-of-and-its-preference-over-i